Create DASG Pull Request Template

karlmdavis opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
N/A (we don't currently have a Jira project for this kind of DASG admin work)

**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
As a DASG staffer,  I need a common Pull Request template that we use across our projects, so that we have a sufficient and more consistent approach to change management across DASG programs.


### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->

Define default Pull Request template for DASG repos.

Defines the default Pull Request template for this repository. The larger intent, though, is that other DASG programs/repos will also adopt this PR template or one derived from it. Its contents are intended to help PR submitters and reviewers:

1. Help ensure that everyone clearly understands not just _what_ is being changed, but also _why_.
2. Help everyone understand what needs to be reviewed in each PR.
3. Help ensure compliance with our standards and processes.

Note: This template will not in any way be "final" or otherwise set in stone. I've heard that BCDA and other programs have regularly iterated on their existing PR templates and I'd hope to see the same happen on this one. Once this PR is merged and rolled out to the other programs, we'll hopefully see many future PRs to improve it.


### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->

If you're reviewing this PR, please check these things, in particular:

* Does this PR template cover the things that your project cares about?
* Will this PR template help submitters and reviewers meet our Engineering Standards?
* Is there anything too "extra" here that we can/should remove, to help out our PR submitters?
* Are there any other helpful comments you think should be added, to help out our PR submitters?


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of these questions is **Yes**, then here's a link to the associated Security Impact Assessment: <https://confluence.cms.gov/display/ODI/SIA+2020-04-21+DASG+Pull+Request+Template>.
    * Does this PR add any new software dependencies? **Yes** or **No**. **No.**
    * Does this PR modify or invalidate any of our security controls? **Yes** or **No**. **Yes**.
    * Does this PR store or transmit data that was not stored or transmitted before? **Yes** or **No**. **No.**
* If the answer to any of these questions is **Yes**, then please add @StewGoin as a reviewer,
  and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **Yes** or **No**. **Yes.**

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: <https://confluence.cms.gov/display/ODI/SIA+2020-04-21+DASG+Pull+Request+Template>.
    * Does this PR add any new software dependencies? **No**.
    * Does this PR modify or invalidate any of our security controls? **Yes**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **Yes**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

* N/A


### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [N/A] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

karlmdavis committed:
Added first draft of a DASG PR template.

This is intended to make our processes and standards clearer for both PR
reviewers and submitters.

karlmdavis committed:
Removed extra line breaks in PR template.

Seems like GitHub's issue/PR descriptions don't quite follow the
markdown spec: single line breaks and extra spaces are not ignored. Oh
well.

msnook commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+**JIRA Ticket:**
+<https://jira.cms.gov/browse/xxxx>
+
+**User Story or Bug Summary:**
```
Can this section be labeled "Description" instead?

msnook commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+    * Does this PR add any new software dependencies? **Yes** or **No**.
+    * Does this PR modify or invalidate any of our security controls? **Yes** or **No**.
+    * Does this PR store or transmit data that was not stored or transmitted before? **Yes** or **No**.
+* If the answer to any of these questions is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
```
This seems like a pretty heavy burden on @StewGoin.  Not to mention, is it a blocker if @StewGoin is out or on vacation?
BCDA has a security checklist process which we go through _before_ doing any work, and get it approved by @StewGoin.  Is this new process intended to deprecate that?

karlmdavis replied:
I think you're reading it backwards? Jason only needs to get tagged for a "Yes" to the question _below_ that.

Or am I misunderstanding you?

dhgreene replied:
True, though it takes at least a second pass to be sure.  How about:
"If the answer to the below question is . . ."

msnook replied:
Ahhh, yes I think you are right.  If I am _now_ understanding you correctly, Jason only gets tagged for this question:
```Do you think this PR requires additional review of its security implications for other reasons?```

I definitely thought it applied to the above section, likely due to the raw format of the file under review.  I suspect, that once properly formatted with markdown, there will be no confusion here.  

Thanks for clarifying.

karlmdavis replied:
Any concerns with that, then? I think it's prompting folks to ask themselves, "_should_ I maybe talk to Jason about this first?" Which is a good thing to think about!

And yeah, if he's out, then... 🤷. It can likely wait. And if it can't, then folks are smart enough to ask for help, I think.

msnook replied:
Totally agree.  I've seen lots of folks do this naturally, but having the prompt in the template is definitely worthwhile.

msnook commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+    1. Ensure that any problems it causes have a small "blast radius".
+    2. Ensure that it'll be easier to rollback if that becomes necessary.
+    3. Ease the burden on reviewers.
+* [ ] I've verified that the individual commits in this PR all:
```
Does this matter?  If we are using "Squash and Merge", then this won't matter, right?  
I'm worried that this will be hard to follow, specifically in situations where a heavy dose of troubleshooting is occurring, and sometimes the commit comments are simply "fixing" or "reverting" (for example).

karlmdavis replied:
I think we have an item in the Engineering Standards right now asking folks to _not_ squash.

Personally, I'm a big fan of a rich & useful history. I **like** knowing how the code evolved as it was drafted: that this line had a bug that was fixed, that that other function was renamed to be more clear, etc. I also think it's good just to be honest about how code doesn't just pop into the world fully-formed and beautiful.

However. That's just a personal opinion, and not a strong enough one that it really rises to the level of needing to be a DASG-wide standard.

Double-however! I think it's useful -- for consistency's sake -- for DASG to have a single merge strategy that gets used across all the projects. Makes it easier for folks to play in other projects.

What are your thoughts?

msnook replied:
I'm a big fan of `Squash and Merge` as it keeps the commit history in `master` clean.  If digging is necessary, the PR from which code was added can be tracked down, and then the commit history can be viewed there.  Personal preference, I guess.

dhgreene replied:
Using squash and merge would also care for the concerns in the previous checkbox.

karlmdavis replied:
@msnook, et al. I've tweaked the wording here to allow for programs that opt to use squash and merge. I want to mention _something_ along these lines still, as I think the squash and merge commit message starts off as a big combo of all the other commit messages, right?

Open to suggestions for better wording here!

msnook commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+* [ ] I've verified that the individual commits in this PR all:
+    1. Reasonably explain the "what" and "why" of the changes.
+    2. Leverage JIRA's [smart commits](https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html) to reference the JIRA ticket that they're associated with.
+* [ ] I've verified that this PR includes all required documentation changes, including changelog / release notes entries.
```
Release note entries?  In BCDA's current process, the PR is self-documenting... in other words, the release notes are generated from the list of PRs included in the release.

dkrylovsb replied:
True for AB2D as well (we're using `gren`).

karlmdavis replied:
Then you're okay?

DeirdreHolub replied:
Can we please add something about README updates consistent with the change? How about `this PR includes all required documentation changes, including README updates and changelog / release notes entries.`

karlmdavis replied:
Excellent suggestion, Deirdre!

Also, as a complete aside, are you familiar with GitHub's new "suggestion" feature? See #6 here: <https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/reviewing-proposed-changes-in-a-pull-request>. Also worth mentioning that GH also now supports multi-line comments and suggestions: you click and drag on the "+" instead of just single-clicking. (Apologies if you were already familiar with these -- just trying to spread the good news!)

dkrylovsb commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+### Submitter Checklist
+
+Before requesting final review, I've gone through and verified that this PR complies with the following requirements:
+
```
I'd add another checkbox, which we have in our templates in AB2D, along these lines: "I have verified that tests/code quality checks pass locally against the branch". Even though PRs are going to get checked automatically by CI tools, having a checkbox like this one would help deter folks from submitting changes they haven't bothered running tests on. Just my two cents.

karlmdavis replied:
Yeah, I saw that one in your template and thought about including it.

I think, though, that's not worth having strong opinions about: I myself prefer to run things locally but I've worked with lots of folks who really do prefer to let the CI do it for them. Different strokes for different folks, I think?

karlmdavis committed:
Tweaked wording for clarity.

Hat tip to @msnook for identifying the need.

karlmdavis commented:
Also adding @johnfrenchxyz as a reviewer. John, would you mind reviewing this from a human usability perspective?

johnfrenchxyz commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+
+Before requesting final review, I've gone through and verified that this PR complies with the following requirements:
+
+* [ ] I've verified that this PR is as small as it can reasonably be, in order to:
```
What would you want a developer to do if they realize, "Oh no, this is definitely too big of a PR!"? Is there a standard or best-practice for refining the scope of a PR that we could point to? Basically, how we'd like them to dig themselves out of the hole?

karlmdavis replied:
Great suggestion!

I had trouble finding articles that were 100% on-topic for this, but I just tweaked the wording and added a couple of references that I think get the job done in concert.

Better?

em1 replied:
Seconding @johnfrenchxyz 's question. I'm not sure there's a clean way to recover if you've reached the point of creating a large PR. This is something we need to be mindful of when creating tickets and while work is being done. Are you making more changes than you expected? Can this work be broken into smaller tasks?

johnfrenchxyz replied:
I think your update gives more context. And I agree that this is more of a pre-PR challenge in ticket creation, grooming, etc. I live having the check from a culture standpoint. I think I am good with the checkbox now if we push to make this a priority in ticket creation.

karlmdavis replied:
FWIW, you can break up a branch/PR after it's gotten too large. It's not a lot of fun, but I've done it.

That said, both of you are **very** correct that it's better to try and prevent than to fix! I'll leave the wording here, as updated, for now but leave a mental note that this might need to be revisited in the future. (Don't think I can add a comment to the Markdown in the middle of the list, or I'd leave a note there for submitters. 🤷)

dhgreene replied:
If we're intending prevention rather than fixing, I'm not sure the PR template is the right place for it.  We risk having so much in here that no one reads/thinks through it all.

sarah-tully commented:
Have we done any usability testing of this template with a few devs within DASG for work that's in progress right now? Before we roll out something to all teams, I'd be curious to understand more about things like:
-  how much longer does this take to fill out than to submit a PR currently?
- Do devs find it easy to fill out? why/why not?
- is there any portion that's particularly challenging to fill out on their own? 
- also, what is the value we hope to gain by implementing it?

I'm happy to help run a quick session if folks think that'd be helpful, if we haven't already.

Also, any thoughts on 'piloting' the first iteration with one team before rolling out to everyone?

DeirdreHolub commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+-->
+
+**JIRA Ticket:**
+<https://jira.cms.gov/browse/xxxx>
```
[PROJ-xxx](https://jira.cms.gov/browse/PROJ-xxx)
`[PROJ-xxx](https://jira.cms.gov/browse/PROJ-xxx)`
will auto format the link nicely, where PROJ is the JIRA project code and xxx is replaced with the ticket #

karlmdavis replied:
I would normally agree with you, but here I _think_ it makes more sense to optimize for speed in filling out the template (by going with just the bare link). What do you think?

DeirdreHolub replied:
I think you click the link more often that you type it.

DeirdreHolub commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+    3. Ease the burden on reviewers.
+* [ ] I've verified that the individual commits in this PR all:
+    1. Reasonably explain the "what" and "why" of the changes.
+    2. Leverage JIRA's [smart commits](https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html) to reference the JIRA ticket that they're associated with.
```
Are you referring to linking **to** the JIRA ticket or **from** the JIRA ticket to the github PR? both are possible and your intent here is not clear.

DeirdreHolub commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+
+### What Needs to Be Deployed Before this PR?
+
+This PR cannot be deployed until the following pre-requisite changes have been fully deployed:
```
Are you saying the PR can be merged when it's not deployable? I remember conversations about writing code intelligent enough to know whether it can run safely or putting it behind a feature flag, but not making  master un-deployable until some other thing happens.

If we are going to add verbiage about using feature flags to protect execution of code until a subsequent merge occurs, we should also document the feature flag and/or add a ticket to remove the flag when the required thing has happened.

karlmdavis replied:
I very much agree that feature flags are a best practice we should adopt! I tend to think we're quite a ways off from baking that into our standards and default practices, though. You?

(Also: tweaked the wording here a bit. That better for now?)

karlmdavis replied:
@DeirdreHolub (bump)

DeirdreHolub commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+* [ ] I've verified that this PR includes all required documentation changes, including changelog / release notes entries.
+* [ ] I've verified that all new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers unfamiliar with the project.
+* [ ] I've verified that all tech debt and/or shortcomings introduced by this PR is detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
+* [ ] I've requested review from at least two other engineers on this project (at least one of whom is a senior engineer or owns the relevant component(s) here).
```
Why are we including this here when presumably github will enforce this?

karlmdavis replied:
The "senior" part can't be automatically enforced, to the best of my knowledge.

dhgreene replied:
The "senior" part is also not something we usually emphasize on our teams generally.

DeirdreHolub commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,84 @@
+    1. Reasonably explain the "what" and "why" of the changes.
+    2. Leverage JIRA's [smart commits](https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html) to reference the JIRA ticket that they're associated with.
+* [ ] I've verified that this PR includes all required documentation changes, including changelog / release notes entries.
+* [ ] I've verified that all new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers unfamiliar with the project.
```
There will be code that simply will not be reasonably clear to engineers unfamiliar with the project, typically because the code requires significant domain knowledge or significant understanding of an uncommon framework. How about `reasonably clear to engineers, preferably ones unfamiliar with the project.`

karlmdavis commented:
> Have we done any usability testing of this template with a few devs within DASG for work that's in progress right now? Before we roll out something to all teams, I'd be curious to understand more about things like:
> 
> * how much longer does this take to fill out than to submit a PR currently?
> * Do devs find it easy to fill out? why/why not?
> * is there any portion that's particularly challenging to fill out on their own?
> * also, what is the value we hope to gain by implementing it?
> 
> I'm happy to help run a quick session if folks think that'd be helpful, if we haven't already.
> 
> Also, any thoughts on 'piloting' the first iteration with one team before rolling out to everyone?

@sarah-tully I'm testing it out right here (see the PR description above). That said, I think it's an excellent idea to test this out some more! In the interests of maximizing ROI on that, how about we get one of the teams to dogfood it for a sprint? We can then incorporate any feedback into a new PR and _then_ roll it out to all of the programs. Would you be willing to have the BFD team do that?

sarah-tully commented:
> > Have we done any usability testing of this template with a few devs within DASG for work that's in progress right now? Before we roll out something to all teams, I'd be curious to understand more about things like:
> > 
> > * how much longer does this take to fill out than to submit a PR currently?
> > * Do devs find it easy to fill out? why/why not?
> > * is there any portion that's particularly challenging to fill out on their own?
> > * also, what is the value we hope to gain by implementing it?
> > 
> > I'm happy to help run a quick session if folks think that'd be helpful, if we haven't already.
> > Also, any thoughts on 'piloting' the first iteration with one team before rolling out to everyone?
> 
> @sarah-tully I'm testing it out right here (see the PR description above). That said, I think it's an excellent idea to test this out some more! In the interests of maximizing ROI on that, how about we get one of the teams to dogfood it for a sprint? We can then incorporate any feedback into a new PR and _then_ roll it out to all of the programs. Would you be willing to have the BFD team do that?

I talked to @arouth and she is going to talk to BCDA about being the guinea pig next sprint to get some feedback and iterate before releasing to the rest of DASG.

msnook commented:
> > > Have we done any usability testing of this template with a few devs within DASG for work that's in progress right now? Before we roll out something to all teams, I'd be curious to understand more about things like:
> > > 
> > > * how much longer does this take to fill out than to submit a PR currently?
> > > * Do devs find it easy to fill out? why/why not?
> > > * is there any portion that's particularly challenging to fill out on their own?
> > > * also, what is the value we hope to gain by implementing it?
> > > 
> > > I'm happy to help run a quick session if folks think that'd be helpful, if we haven't already.
> > > Also, any thoughts on 'piloting' the first iteration with one team before rolling out to everyone?
> > 
> > 
> > @sarah-tully I'm testing it out right here (see the PR description above). That said, I think it's an excellent idea to test this out some more! In the interests of maximizing ROI on that, how about we get one of the teams to dogfood it for a sprint? We can then incorporate any feedback into a new PR and _then_ roll it out to all of the programs. Would you be willing to have the BFD team do that?
> 
> I talked to @arouth and she is going to talk to BCDA about being the guinea pig next sprint to get some feedback and iterate before releasing to the rest of DASG.

@arouth @sarah-tully We have been refining our PR template over the last year or so, and usability is fine.  Also, if any issues in usability pop up, folks either bring it up in a retro OR create a PR to fix the issues (as evident in a refinement we made to our template just this sprint!). Also, our version is quite similar to this version.  I don't know if you'd gain a lot of feedback using BCDA as a guinea pig.

karlmdavis committed:
Mentioned README updates.

Hat tip to @DeirdreHolub for the suggestion!

karlmdavis committed:
Change the Jira link formatting.

Now optimizes for readability at the cost of a bit of extra work for PR
submitters. Hat tip to @DeirdreHolub for the suggestion!

karlmdavis committed:
Better explain Jira's smart commits with example.

Now also provides an incidental link to general commit formatting
guidelines. Hat tip to @DeirdreHolub for the suggestion!

karlmdavis committed:
Tweaked the prereq wording to be a bit clearer.

Unless/until we have feature flags as a widespread practice, folks
should generally not be merging things that can't be safely deployed.

Hat tip to @DeirdreHolub for the call-out.

karlmdavis committed:
Tweaked wording on code comments.

Hat tip to @DeirdreHolub for the suggestion!

karlmdavis committed:
Add some doc links on PR fixup.

I couldn't find any articles that were 100% spot-on topic, but the combo
of these two articles gets close enough, I think.

Hat tip to @johnfrenchxyz for the suggestion!

dhgreene commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+
+Submitters should complete the following questionnaire:
+
+* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment: N/A.
```
There's more nuance here.  In BCDA, we have a security checklist that helps us determine whether an SIA is needed (it asks these questions and more); linking to the security checklist would be an appropriate substitute if a SIA was _not_ needed.  

In DPC, we follow a similar but less formal process.   Our understanding is that context matters; for instance, a non-security-related dependency that is limited to development or CI/CD should not trigger a SIA.  If we're not sure, we do reach out to @StewGoin, but almost always in a more private forum than the public PR review.

This may just be something each team has to tweak to fit their security process.

karlmdavis replied:
@dhgreene If we can at all manage it, I'd prefer to avoid having teams customize their templates, as that will make change management a mess -- it'll be a lot harder to roll out changes to this template if everyone's customized it.

Can you suggest any wording changes/additions here that would make it a better fit for BCDA & DPC?

Also worth noting that this language was cleared with @StewGoin ahead of this PR going out. That's not to say it can't be improved! Merely pointing out that it's at least _sufficient_ from his perspective.

dhgreene replied:
```suggestion
* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA) (or security checklist documenting why a SIA is not needed): N/A.
```

karlmdavis replied:
@dhgreene What about this, instead?
```suggestion
* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
```

em1 commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+* [ ] I've verified that this PR includes all required documentation changes, including `README` updates and changelog / release notes entries.
+* [ ] I've verified that all new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
+* [ ] I've verified that all tech debt and/or shortcomings introduced by this PR is detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
+* [ ] I've requested review from at least two other engineers on this project (at least one of whom is a senior engineer or owns the relevant component(s) here).
```
```suggestion
* [ ] I've requested review from at least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
```
This looks like a suggestion and not a requirement if it's in parentheses. But maybe it is a suggestion...?

karlmdavis committed:
Fix punctuation on review requirements.

Hat tip to @embh for the suggestion!

Co-Authored-By: Emily Hart <1923441+embh@users.noreply.github.com>

orenfromberg reviewed (commented):
This is a great PR template, if not a guide on how to author an PR, that is exhaustive in terms of addressing common issues that often manifest on inexperienced teams (and occasionally experienced ones!). It is opinionated and prescriptive with respect to the way a team reviews and merges code, as it should be. I think this would provide a solid resource for individual teams to reference and customize accordingly in an ad hoc manner with respect to the common issues that are regularly addressed within the scope of their repositories.

dhgreene commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+Add some items to the list above, or remove the entire section if it doesn't apply.
+
+Common items include:
+* Database migrations (which should always go out in PRs by themselves, to reduce risk).
```
```suggestion
* Database migrations (which should always be deployed by themselves, to reduce risk).
```

karlmdavis committed:
Update guidance to allow for squash and merge.

karlmdavis committed:
Merge branch 'pull-request-template' of github.com:CMSgov/cms-oeda-dasg into pull-request-template

karlmdavis committed:
Update .github/pull_request_template.md

Co-Authored-By: dhgreene <dhgreene@gmail.com>

karlmdavis commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+* [ ] I've verified that this PR includes all required documentation changes, including `README` updates and changelog / release notes entries.
+* [ ] I've verified that all new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
+* [ ] I've verified that all tech debt and/or shortcomings introduced by this PR is detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
+* [ ] I've requested review from at least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
```
@DeirdreHolub and @dhgreene I'm thinking of striking this item entirely and instead asking each team's tech lead to use GitHub's "code owners" and branch protection features to designate a list of "approving engineers" for the whole repository.

Thoughts?

For reference: https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners

karlmdavis replied:
Whoops, that's the wrong link. Fixed it.

karlmdavis replied:
Going to hold off on this change for now, as getting that deployed everywhere will be a project. I still want to do it (always better to automate stuff like that!) but I'm kicking the can down the road for a bit.

karlmdavis commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+* [ ] I've verified that all tech debt and/or shortcomings introduced by this PR is detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
+* [ ] I've requested review from at least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
+* [ ] I've requested review from any relevant engineers on other projects (e.g. BFD, SLS, etc.).
+* [ ] I've verified that this PR and its changes comply with all other policies in the [DASG Engineering Standards](../policies/engineering_standards.md) and have specifically called out any/all deviations in this PR.
```
@DeirdreHolub Does this address the concern you'd mentioned elsewhere?

```suggestion
* [ ] I've verified that this PR and its changes comply with all other policies in the [DASG Engineering Standards](../policies/engineering_standards.md) and have specifically called out any/all deviations in this PR.
    * If you haven't reviewed the standards in the last few months, please do!
```

DeirdreHolub replied:
It's the wording of the assertion that is at issue. How about `I've done my best to write this PR and its changes in compliance with all other` and  the of the words.

karlmdavis replied:
I believe my recent changes address this concern.

karlmdavis commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+* [ ] I've verified that all tech debt and/or shortcomings introduced by this PR is detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
+* [ ] I've requested review from at least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
+* [ ] I've requested review from any relevant engineers on other projects (e.g. BFD, SLS, etc.).
+* [ ] I've verified that this PR and its changes comply with all other policies in the [DASG Engineering Standards](../policies/engineering_standards.md) and have specifically called out any/all deviations in this PR.
```
@DeirdreHolub Or perhaps this?
```suggestion
* [ ] I've verified that this PR and its changes comply with all other policies in the [DASG Engineering Standards](../policies/engineering_standards.md) and have specifically called out any/all deviations in this PR.
    * Please review the standards every few months to ensure you're familiar with them.
```

RickHawesUSDS replied:
Why not directly ask PR writers to list deviations from the DASG engineering standard? It's a tone comment. Checklists should be easy to scan. The "I have verified this PR and its changes" statements make theses statement list items hard to scan and easy to automatically check.

karlmdavis replied:
Excellent suggestion! I believe my recent changes address this concern.

RickHawesUSDS commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,96 @@
+    * Does this PR modify or invalidate any of our security controls? **Yes** or **No**.
+    * Does this PR store or transmit data that was not stored or transmitted before? **Yes** or **No**.
+* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
+    * Do you think this PR requires additional review of its security implications for other reasons? **Yes** or **No**.
```
Shouldn't the process be that Jason reviews and approves the SIA? Does he also have to approve the PR?

karlmdavis replied:
Only when a SIA isn't required but the submitter still thinks he should take a look at the PR.

Very open to suggestions on better wording, here.

karlmdavis committed:
Soften wording on docs checklist item.

"Any" rather than "all" allows for judgement calls, as to when docs
are/aren't required.

karlmdavis committed:
Fixed minor grammar issue.

karlmdavis committed:
Tweak SIA wording.

karlmdavis committed:
Adjusted checklist wording, for read-/scan-ability.

Trying to address the excellent feedback from @RickHawesUSDS and
@DeirdreHolub, with this.

karlmdavis committed:
Moved comment blocks to top of their section.

The comments had been at the bottom of the sections, which I think makes
them easier to miss.

karlmdavis committed:
Merge branch 'pull-request-template' of github.com:CMSgov/cms-oeda-dasg into pull-request-template

karlmdavis commented:
The BFD team piloted this PR template a sprint or so ago and engaged with design staff to collect and synthesize feedback from that trial. Feedback is documented here: https://miro.com/app/board/o9J_krV0yFA=/

Overall, it was an excellent exercise (great suggestion, @sarah-tully!) and resulted in some actionable feedback, which has since been incorporated here.

jzulim reviewed (approved):
I've used this PR template several times and participated in the research exercise. It's great to see that feedback incorporated. Thank you for spearheading this. Approve.d

jzulim commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,98 @@
+* [ ] Reviews are requested from both:
+    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
+    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
+* [ ] Any deviations from the other policies in the [DASG Engineering Standards](../policies/engineering_standards.md) are specifically called out in this PR, above.
```
Should we make this path absolute so that when used outside of this repo the link resolves?

karlmdavis replied:
Excellent idea -- done!

karlmdavis committed:
Made Standards link absolute.

Per excellent suggestion from @jzulim, this will ensure it works when
the template is used in other repos.

karlmdavis committed:
Tweak "small blast radius" item's wording.

Hat tip to @dhgreene for a thoughtful discussion on this one!

karlmdavis commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,107 @@
+### What Does This PR Do?
+
+<!--
+Add detailed discussion of changes here
+This is likely a summary, or the complete contents, of your commit messages.
```
```suggestion
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
```

karlmdavis commented on .github/pull_request_template.md:
````diff
@@ -0,0 +1,107 @@
+    2. There isn't too much of a burden on reviewers.
+    3. Any problems it causes have a small "blast radius".
+    4. It'll be easier to rollback if that becomes necessary.
+* [ ] The commits in this PR:
+    1. Reasonably explain the "what" and "why" of the changes.
+    2. Leverage JIRA's [smart commits](https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html) to reference the JIRA ticket that they're associated with. For example, aim for commit messages like this (note also [the 50/72 formatting](https://stackoverflow.com/q/2290016) used here):
+        
+        ```
+        Added the whizbang to the doodad.
+
+        The new whizbang should really simplify things for our whoosit users.
+        It was a bit tricky to get it into the doodad, but we decided that the
+        flipwhee pattern was the best approach for now. Might want to revisit
+        that in the future if it ends up being too hard to maintain.
+
+        SOMEPROJECT-42
+        ```
+        
````
```suggestion
* [ ] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
```

karlmdavis reviewed (commented):
Hey, @dhgreene: thoughts on these two changes, re: commit practices?

karlmdavis committed:
Apply suggestions from code review

Reworks standards on commit messages.

Ensures that final commit message is included in PR, when appropriate, so that it can get reviewed. Also ensures that PRs get linked to Jira. All while allowing for teams that prefer a squash-and-merge approach.

dhgreene reviewed (approved):
Thanks for the flexibility and dialog!  👍

karlmdavis commented:
Very meta of me, but I updated the PR description here to use the latest version of the template (barring any copy-paste goofs, at least). Going to merge it now.

Thanks to everyone for the feedback & suggestions: this is now noticeably better than it started off as. As our 3 year old likes to say: Teamwork makes the dream work!
